### PR TITLE
Implement metrics for c3p0 connection pool

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
CC @theletterf for minor `metrics.md` change

### CODE REVIEW

**Review by breedx-splk — COMMENTED** [review]
Couple small questions/changes, otherwise looks good.

It also occurs to me that we are not tracking high watermark for the pool. I've wanted something like this in the past -- a way to see what the highest connection usage was over time. I suppose the backend/o11y tools can probably answer this (eg. max(active)) but it might also be nice to report it locally?  Ignore me if this has been discussed already.  😁

**Review by theletterf — APPROVED** [review]

File: instrumentation/c3p0/src/test/java/com/splunk/opentelemetry/c3p0/MockDriver.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  @Override
+  public boolean jdbcCompliant() {
+    return false;
+  }
+
+  @Override
+  public Logger getParentLogger() throws SQLFeatureNotSupportedException {
+    throw new SQLFeatureNotSupportedException("Feature not supported");
+  }
+}
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply instrumentation/c3p0/src/test/java/com/splunk/opentelemetry/c3p0/MockDriver.java:69]: Not for `MockDriver`, but I think connection can be mocked -- will try 👍

File: instrumentation/c3p0/src/main/java/com/splunk/opentelemetry/c3p0/AbstractPoolBackedDataSourceInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+import com.mchange.v2.c3p0.impl.AbstractPoolBackedDataSource;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+final class AbstractPoolBackedDataSourceInstrumentation implements TypeInstrumentation {
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply instrumentation/c3p0/src/main/java/com/splunk/opentelemetry/c3p0/AbstractPoolBackedDataSourceInstrumentation.java:28]: Yeah, I used the `<instrumented class name>Instrumentation` naming convention for that... it does look pretty weird here

File: instrumentation/c3p0/src/test/java/com/splunk/opentelemetry/c3p0/C3p0InstrumentationTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            new MeterData("db.pool.connections", "gauge", "connections", tags),
+            new MeterData("db.pool.connections.active", "gauge", "connections", tags),
+            new MeterData("db.pool.connections.idle", "gauge", "connections", tags),
+            new MeterData("db.pool.connections.pending_threads", "gauge", "threads", tags));
+  }
+
+  private static void assertNoConnectionPoolMetrics() {
+    assertThat(TestMetricsAccess.getMeters().stream().map(MeterData::getName).distinct())
+        .doesNotContain(
+            "db.pool.connections",
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply instrumentation/c3p0/src/test/java/com/splunk/opentelemetry/c3p0/C3p0InstrumentationTest.java:86]: Good point, I reused the `DataSourceSemanticConventions` class here

File: instrumentation/c3p0/src/main/java/com/splunk/opentelemetry/c3p0/ConnectionPoolMetrics.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+public final class ConnectionPoolMetrics {
+
+  // a weak map does not make sense here because each Meter holds a reference to the dataSource
+  // DataSourceProxy does not implement equals()/hashCode(), so it's safe to keep them in a plain
+  // ConcurrentHashMap
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply instrumentation/c3p0/src/main/java/com/splunk/opentelemetry/c3p0/ConnectionPoolMetrics.java:43]: Oops, I've just noticed that this comment is not true for this connection pool implementation... the dangers of copy-paste 🥶

File: instrumentation/c3p0/src/test/java/com/splunk/opentelemetry/c3p0/C3p0InstrumentationTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply instrumentation/c3p0/src/test/java/com/splunk/opentelemetry/c3p0/C3p0InstrumentationTest.java]: I'll convert all other connection pool tests to the new pattern in a separate PR.

**Comment by mateuszrzeszutek** [comment]
> It also occurs to me that we are not tracking high watermark for the pool. I've wanted something like this in the past -- a way to see what the highest connection usage was over time. I suppose the backend/o11y tools can probably answer this (eg. max(active)) but it might also be nice to report it locally? Ignore me if this has been discussed already. 😁

I don't think we've discussed that before, but as you said - it's probably trivial to achieve that in the UI. I think I'd prefer sending whatever metrics a particular pool exposes, unless there's a requirement for computing such stuff here.

**Review by breedx-splk — APPROVED** [review]

**Review by iNikem — APPROVED** [review]

**Pull request merged**